onnxruntime/test/optimizer/graph_transform_test.cc

```diff
@@ -541,6 +541,27 @@ TEST(GraphTransformationTests, MatMulAddFusion_three_input) {
   ASSERT_TRUE(op_to_count["Gemm"] == 1);
 }
 
+// Matmul+Add with shape [k]*[k,N]+[N], won't do the fusion
+// We can do the fusion by changing shape to [1,k]*[k,N]+[1,N], then add a reshape [1,N]=>[N]
+// This will bring extra cost. And there's only very limited gain to fuse Matmul+Add to Gemm
+// Since the basic implementation is almost same
+TEST(GraphTransformationTests, MatMulAddFusion_negitive_case) {
+  auto model_uri = MODEL_FOLDER "matmul_add_fusion/3Input/neg_model.onnx";
+
+  std::shared_ptr<Model> p_model;
```

Comment:
>shared_ptr [](start = 7, length = 10)

Do we actually load the model into shared_ptr? Has Yuan seen this?

Reply:
we do have the interface use the shared_ptr I'm not sure the reason. Let's have another PR to clean up it if you have any concern.
[https://github.com/microsoft/onnxruntime/blob/aa7c79eac9875ba3b5c9444b13d1878560415ea7/onnxruntime/core/graph/model.cc#L378](https://github.com/microsoft/onnxruntime/blob/aa7c79eac9875ba3b5c9444b13d1878560415ea7/onnxruntime/core/graph/model.cc#L378)

---
In reply to: [347538251](https://github.com/microsoft/onnxruntime/pull/2407#discussion_r347538251) [](ancestors = 347538251)